### PR TITLE
add py-h5py@3.11

### PR DESCRIPTION
**Pull request by s-sajid-ali**
https://docs.h5py.org/en/stable/whatsnew/3.11.html

### CODE REVIEW

**Comment by takluyver** [comment]
h5py 3.10 and above requires HDF5 1.10.4 or above - I think this change was missed when 3.10 was added.

Versions before 3.11 also won't pass the full test suite with Numpy 2.0 (which is coming out soon), though I think they're mostly compatible in practice. 3.11 is the first version that's been tested with Numpy 2 (RC).

File: var/spack/repos/builtin/packages/py-h5py/package.py [hunk 1]
```diff
@@ -55,15 +58,17 @@ class PyH5py(PythonPackage):
     depends_on("py-setuptools", type="build")
 
     # Build and runtime dependencies
-    depends_on("py-numpy@1.17.3:", type=("build", "run"), when="@3.9:")
-    depends_on("py-numpy@1.19.3:", type=("build", "run"), when="@3:3.5 ^python@3.9.0:")
-    depends_on("py-numpy@1.17.5:", type=("build", "run"), when="@3:3.5 ^python@3.8.0:3.8")
-    depends_on("py-numpy@1.14.5:", type=("build", "run"), when="@3:")
-    depends_on("py-numpy@1.7:", type=("build", "run"), when="@:2")
+    depends_on("py-numpy@1.17.3:", type=("build", "run"), when="@3.11:")
+    depends_on("py-numpy@1.17.3:2", type=("build", "run"), when="@3.9:3.11")
```
Comment (takluyver) [inline var/spack/repos/builtin/packages/py-h5py/package.py]:
```suggestion
    depends_on("py-numpy@1.17.3:1", type=("build", "run"), when="@3.9:3.11")
```

Double check this, I may have misunderstood how Spack's version ranges work, but looking at other ones in this file, I think they're inclusive, so `:1` allows any 1.x.y.

File: var/spack/repos/builtin/packages/py-h5py/package.py [hunk 2]
```diff
@@ -55,15 +58,17 @@ class PyH5py(PythonPackage):
...
+    depends_on("py-numpy@1.19.3:2", type=("build", "run"), when="@3:3.5 ^python@3.9.0:")
+    depends_on("py-numpy@1.17.5:2", type=("build", "run"), when="@3:3.5 ^python@3.8.0:3.8")
+    depends_on("py-numpy@1.14.5:2", type=("build", "run"), when="@3:")
+    depends_on("py-numpy@1.7:2", type=("build", "run"), when="@:2")
 
     # Link dependencies (py-h5py v2 cannot build against HDF5 1.12 regardless
     # of API setting)
-    depends_on("hdf5@1.8.4:1.14 +hl", when="@3.8:")
+    depends_on("hdf5@1.10.4:1.14 +hl", when="@3.10:")
+    depends_on("hdf5@1.8.4:1.14 +hl", when="@3.8:3.10")
```
Comment (takluyver) [inline var/spack/repos/builtin/packages/py-h5py/package.py]:
```suggestion
    depends_on("hdf5@1.8.4:1.14 +hl", when="@3.8:3.9")
```

Ditto

File: var/spack/repos/builtin/packages/py-h5py/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: That is correct (see https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges).

**Review by tldahlgren — COMMENTED** [review]
Confirmed the version sha256, agree with comment about version range checks.

File: var/spack/repos/builtin/packages/py-h5py/package.py (hunk 1 above)
Comment (s-sajid-ali) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Fixed. Thanks for catching this!

File: var/spack/repos/builtin/packages/py-h5py/package.py (hunk 2 above)
Comment (s-sajid-ali) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Fixed.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**